envi/archs/rx/disasm.py

```diff
@@ -0,0 +1,1349 @@
+                        opsize += 1
+                        opvals[opsize] = val
+                        #print(" ++ %x  0x%x  %r" % (bval, val, opvals))
+
+
+                    # find a match:
+                    if opvals[opsz] & mask != endval:
+                        #masked = opvals[opsz] & mask
+                        #print("-cmp- 0x%x & 0x%x (0x%x) != 0x%x" % (opvals[opsz], mask, masked, endval))
+                        continue
+
+                    found = True
+                    break
+
+                except Exception as e:
+                    # if we don't have enough bytes to grow to a certain size, that possible decoding is invalid
+                    #print("%r (%r)" % (e, type(e)))
+                    continue
+
+        else:   # true up the length for maintable entries
+            while opsize < opsz:
+                # grab the next byte
+                bval, = struct.unpack_from('B', bytez, off+opsize)
+                # update the running value we'll use to parse
+                val <<= 8
+                val |= bval
+                # update the opsize and store the byte
+                opsize += 1
+                opvals[opsize] = val
+
+
+        if not found:
+            raise e_exc.InvalidInstruction(bytez[off:off+opsize], "Couldn't find a opcode match in the table")
+
+        logger.debug("PARSE MATCH FOUND: val: %x\n\t%x %x %x %r %r  %r  %x", val, mask, endval, opcode, mnem, operdefs, opsz, iflags)
+
+
+        # we've found a match, parse it!
+        off += opsz
+        val = opvals[opsz]
+
+        # let's parse out the things
+        fields = {}
+        for fkey, fparts in operdefs:
+            fdata = 0
+            fullmask = 0
+
+            for shval, fmask in fparts:
+                fullmask |= fmask
+                fdata |= ((val >> shval) & fmask)
+
+            if fullmask > 0xff:
+                # this is a >8 bit, we need to reverse byte order
+                isize = 0
+                newdata = 0
+                while fullmask:
+                    isize += 1
+                    newdata <<= 8
+                    newdata |= fdata & 0xff
+                    fdata >>= 8
+                    fullmask >>= 8
+
+                if fkey != O_UIMM and 1 < isize < 4:
+                    fdata = e_bits.sign_extend(newdata, isize, 4)
+                    #print("0x%x: signed: %d, %d, 0x%x -> 0x%x" % (va, isize, 4, newdata, fdata))
+                else:
+                    #print("fkey: %r: %r" % (fkey, nms[fkey]))
+                    fdata = newdata
+                    #print("0x%x: NOT signed: %d, %d, 0x%x -> 0x%x" % (va, isize, 4, newdata, fdata))
+
+            elif fullmask == 0xff:
+                if fkey != O_UIMM:
+                    fdata = e_bits.sign_extend(fdata, 1, 4)
+            
+            fields[fkey] = fdata
+
+        if handler is not None:
+            # if we have a handler, just let it do everything
+            hndlFunc = self.HANDLERS[handler]
```

Comment:
Since each of the sub-functions calls into e_bits.parsebytes, can we wrap this into a try/catch and raise proper exceptions?